helix-term/src/commands.rs

```diff
@@ -5100,3 +5102,240 @@ fn rename_symbol(cx: &mut Context) {
+            };
+
+            let mut new_length = replacement.chars().count();
+
+            let old_length_no_sign = number_text.chars().count();
+            if new_length < old_length_no_sign && (radix != 10 || number_text.starts_with('0')) {
+                replacement = "0".repeat(old_length_no_sign - new_length) + &replacement;
+                new_length = old_length_no_sign;
+            }
+
+            if radix == 10 && new_value.is_negative() {
+                replacement = format!("-{}", replacement);
+                new_length += 1;
+            }
+
+            (replacement, new_length)
+        };
+
+        let changes = std::iter::once((start, end, Some(replacement.into())));
+        let mut transaction = Transaction::change(text, changes);
+
+        // Move cursor to the last character of the number.
+        let mut selection = doc.selection(view.id).clone();
+        let mut primary = selection.primary_mut();
+        primary.anchor = if new_length < old_length {
+            end - 1 - old_length + new_length
+        } else {
+            end - 1 + new_length - old_length
+        };
+        primary.head = primary.anchor;
+        transaction = transaction.with_selection(selection);
+
+        doc.apply(&transaction, view.id);
+        doc.append_changes_to_history(view.id);
+    }
+}
+
+struct NumberInfo {
+    start: usize,
+    end: usize,
+    value: i64,
+    radix: u32,
+}
+
+/// If there is a number at `char_idx`, return the text range, value and radix.
+fn number_at(text: &Rope, char_idx: usize) -> Option<NumberInfo> {
+    let line = text.char_to_line(char_idx);
+    let line_start = text.line_to_char(line);
+    let line = text.line(line);
+    let line_len = line.len_chars();
+
+    let mut pos = char_idx - line_start;
+    let mut range_and_radix = None;
+
+    // Search from the cursor until a number is found or we reach the end of the line.
+    while range_and_radix.is_none() && pos < line_len {
+        pos += line.chars_at(pos).take_while(|c| !c.is_digit(16)).count();
+
+        range_and_radix = if let Some((start, end)) = hex_number_range(&line, pos) {
+            Some((start, end, 16))
+        } else if let Some((start, end)) = octal_number_range(&line, pos) {
+            Some((start, end, 8))
+        } else if let Some((start, end)) = binary_number_range(&line, pos) {
+            Some((start, end, 2))
+        } else if let Some((start, end)) = decimal_number_range(&line, pos) {
+            // We don't want to treat the '0' of the prefixes "0x", "0o", and "0b" as a number itself, so check for that here.
+            if end - start == 1
+                && line.char(start) == '0'
+                && start + 2 < line_len
+                && ((line.char(start + 1) == 'x' && line.char(start + 2).is_digit(16))
+                    || (line.char(start + 1) == 'o' && line.char(start + 2).is_digit(8))
+                    || (line.char(start + 1) == 'b' && line.char(start + 2).is_digit(2)))
```

Comment:
more efficiently you could

```rust
matches!(Cow::from(line.slice(start..start+1)), "0x" | "0o | "0b")
```

Reply:
Yeah, that's much better!

Reply:
Actually this wouldn't work because we also need to check the character after the prefix to see if it's a digit of the correct base. If it's not, the `0` should be treated as a decimal.